grpc/runtime/src/main/scala/io/buoyant/grpc/runtime/DecodingStream.scala

```diff
@@ -0,0 +1,288 @@
+          val doRelease = () => { p.setDone(); Future.Unit }
+          rest -> doRelease
+        }
+    }
+
+    private[this] def mk(frame: h2.Frame): Releaser = frame match {
+      case f: h2.Frame.Data => FrameReleaser(f.buf.length, f.release, Empty)
+      case f: h2.Frame.Trailers => FrameReleaser(0, f.release, Empty)
+    }
+  }
+
+  private[this] case class Header(compressed: Boolean, size: Int)
+
+  private[this] val GrpcFrameHeaderSz = Codec.GrpcFrameHeaderSz
+
+  private val hasHeader: ByteBuffer => Boolean =
+    bb => { GrpcFrameHeaderSz <= bb.remaining }
+
+  private val hasMessage: (ByteBuffer, Int) => Boolean =
+    (bb, msgsz) => { msgsz <= bb.remaining }
+
+  private def decodeHeader(bb0: ByteBuffer): Header = {
+    require(hasHeader(bb0))
+    val bb = bb0.duplicate()
+    bb.limit(bb.position + GrpcFrameHeaderSz)
+    val compressed = (bb.get == 1)
+    val sz = bb.getInt
+    Header(compressed, sz)
+  }
+
+  private def decode[T](
+    s: RecvState,
+    decoder: ByteBuffer => T
+  ): (RecvState, Option[Stream.Releasable[T]]) =
+    s match {
+      case RecvState.Buffer(Some(h0), buf, releaser) =>
+        val Buf.ByteBuffer.Owned(bb0) = Buf.ByteBuffer.coerce(buf)
+        decodeMessage(h0, bb0.duplicate(), releaser, decoder)
+
+      case s => (s, None)
+    }
+
+  private def decodeFrame[T](
+    s0: RecvState,
+    frame: h2.Frame.Data,
+    decoder: ByteBuffer => T
+  ): (RecvState, Option[Stream.Releasable[T]]) =
+    s0 match {
+      case RecvState.Buffer(None, initbuf, releaser0) =>
+        val buf = initbuf.concat(frame.buf)
+        val Buf.ByteBuffer.Owned(bb0) = Buf.ByteBuffer.coerce(buf)
+        val bb = bb0.duplicate()
+        val releaser = releaser0.track(frame)
+
+        if (hasHeader(bb)) {
+          val h0 = decodeHeader(bb)
+          bb.position(bb.position + GrpcFrameHeaderSz)
+          decodeMessage(h0, bb, releaser.consume(GrpcFrameHeaderSz), decoder)
+        } else (RecvState.Buffer(None, buf, releaser), None)
+
+      case RecvState.Buffer(Some(h0), initbuf, releaser) =>
+        val buf = initbuf.concat(frame.buf)
+        val Buf.ByteBuffer.Owned(bb0) = Buf.ByteBuffer.coerce(buf)
+        decodeMessage(h0, bb0.duplicate(), releaser.track(frame), decoder)
+    }
+
+  private def decodeMessage[T](
+    h0: Header,
+    bb: ByteBuffer,
+    releaser: Releaser,
+    decoder: ByteBuffer => T
+  ): (RecvState, Option[Stream.Releasable[T]]) =
+    if (h0.compressed) throw new IllegalArgumentException("compression not supported yet")
+    else if (hasMessage(bb, h0.size)) {
+      // Frame fully buffered.
+      val end = bb.position + h0.size
+      val (nextReleaser, release) = releaser.consume(h0.size).releasable()
+      val msg = {
+        val msgbb = bb.duplicate()
+        msgbb.limit(end)
+        val msg = decoder(msgbb)
+        Some(Stream.Releasable(msg, release))
+      }
+      bb.position(end)
+      if (hasHeader(bb)) {
+        // And another header buffered...
```

Comment:
There may be multiple messages in the frame, yes.  We could introduce a new RecvState that can hold additional fully-parsed messages, but I think it simpler to just read the next header so that on the next call to recv(), the next message will be read and returned.